### ISSUE TITLE
transfomer of v2 got a nan with multi-gpu

### ISSUE THREAD

**Issue by venuswu** [tensorflow/models#7557]
I ran the latest official/models/transformer with tf-nightly-gpu-2.0-preview=20190913 on 8 gpus, the nan occurs when it ran nearly 100 steps. But it ran normally on a single gpu.

Is there something I should install or to configure ??

**Comment by saberkun** [comment]
Thank you for reporting.
Could you share us your command line? Did you find this when you run transformer inside v2 folder or not?

**Comment by venuswu** [comment]
I send you an email. @saberkun

**Comment by venuswu** [comment]
@saberkun 
I ran transformer of v2 with  tf-nightly-gpu (1.15.0.dev20190730) on 8gpus,  the distributed strategy seems invalid, because I compared the figure of loss with an older official-transformer trained with tf1.11 on 8gpu, which seems converge much faster,

when It reaches 18,000 steps, the loss is below 1.10, but it is still above 1.2 for the newest transformer of v2 trained with tf-nightly-gpu (1.15.0.dev20190730) on 8gpus when it reaches 60,000 steps. I verified on a single gpu to train transformer v2, the figure of loss seems much like the figure trained on 8gpus. So I assume the distributed strategy of 8gpu is invalid, the transfomer of v2 is still trained on a single gpu.

So I change the version of tensorflow to the latest tf-nightly-gpu-2.0-preview(2.0.0.dev20190913​) with python3.6, which seems even worse on 8gpus.

The loss ran into nan nearly 200 steps, but it went really well on a single gpu.

the params I ran is as follow:

--data_dir=$DATA_DIR --model_dir=$MODEL_DIR \
    --vocab_file=$VOCAB_FILE --param_set=$PARAM_SET \
    --train_steps=$TRAIN_STEP --steps_between_evals=$TRAIN_STEP \
    --batch_size=4096 --max_length=256 \
    --num_gpus=8 \
    --dtype=fp32 \
    --enable_time_history=False \
    --enable_xla=False \
    --enable_grappler_layout_optimizer=True

I ran this command inside the v2 folder, like this.

export PYTHONPATH="$PYTHONPATH:$HOME/git/models"

cd /home/git/models/official/transformer/v2
My training is processed on a local host V100.

**Comment by saberkun** [comment]
Can you consider to increase your batch size? batch_size = 4096*8
You have 8 gpus and the per core batch size is 4096.

**Comment by venuswu** [comment]
It seems to go well with batchsize = 4096*8 without any nan when running to 3000 steps, 
but what does the param batch_size mean? in early version like tf1.11, it seams every replica for a single gpu can get can get a batch of size 4096. I have noticed in the v2/transformer_benchmark.py, the param has increased to 8 times, and also in the file like transfomer_main.py, when it comes to tpu, the batch_size has to be divided by the number of gpu, but nothing to do with other devices. So what does the param batch_size mean now ? @saberkun

**Comment by saberkun** [comment]
It is the global batch size. tf 1.11 is using ps-worker which is different for sync training I feel. 4096 per replica means the number of tokens per batch for a single core.
There are some explicit handling with TPU (which is not a stable API yet).

**Comment by venuswu** [comment]
In tf1.11, it was already tf.contrib.distribute.MirroredStrategy. @saberkun

**Comment by saberkun** [comment]
Sorry, I feel I made you confused. I mean estimator version.
Never mind, just be clear that the batch size is global batch size and flags should be consistent with https://github.com/tensorflow/models/blob/master/official/transformer/v2/transformer_benchmark.py#L132
These are the settings we tested.

@guptapriya Priya can have a better explanation for v1/v2 mirrorstrategy. I remember the implementation details are different.

**Comment by guptapriya** [comment]
Yes batch_size is set to global batch size for transformer. @saberkun -
maybe good to clarify in the README
<https://github.com/tensorflow/models/blob/master/official/transformer/v2/README.md>
?

Distribution strategy itself doesn't have a notion of batch size, so it is
more about the model code / actual program and how it is set there..

**Comment by saberkun** [comment]
Yes, will do. I will update readme and flag description. It is related how do we set input pipelines.
